architecture/design/multi-region-2DC-deployment.md

```diff
@@ -216,13 +216,61 @@ Upon receiving a `ReplicateMsg` that is a part of a multi-row transaction, the *
 
 This combination of `WAITING_TO_COMMIT` status and `tablets_received_applying_record` field will be used to ensure that transaction records are applied atomically on the replicated universe.
 
+# Handling network partitions and consumer failures
+
+As mentioned before, we replicate records from the WAL (internally known as log) only, so if WAL files containing unreplicated records are garbage collected because of all of its records have already been flushed, these records won't be able to be replicated. So when a network partition happens between two universes being replicated (2DC), or when the consumer data center fails, we want to keep the logs in the producer as long as possible so that once the network or the consumer data center recovers, we can continue replicating the changes to the consumer using the normal replication mechanism (no bulk copy of data required).
+
+## Current log retention policies
+
+Currently, we have two different policies for deciding which log segments to retain. A time-based policy and a policy based on the minimum number of files. These policies are controlled by the gflags `log_min_seconds_to_retain` with a default value of `300` and `log_min_segments_to_retain` with a default value of `2`.
+
+For the time-based policy, the Log object also uses a private variable `wal_retention_secs_` which gets modified through an `ALTER TABLE` request. When deciding if a log file is old enough to be deleted, we use `max(log_min_seconds_to_retain, wal_retention_secs_)`.
+
+To avoid deleting log segments that contain unreplicated entries, we will add a new policy based on the minimum op id that we want to retain. This new policy will be disabled by default and will be controlled by the gflag `enable_log_retention_by_op_idx`.
+
+Because of the possibility of very long network partitions or unused cdc streams that were not deleted, we will have an option in which the op id policy is ignored if the tserver is running out of disk space. If the gflag `log_stop_retaining_min_disk_mb` is set, we will stop retaining records with unreplicated entries once disk space is less than the specified value. This gflag will not affect log segments that are being retained because its entries haven’t been flushed to rocksDB.
```

Comment:
```suggestion
Because of the possibility of very long network partitions or unused cdc streams that were not deleted, we will have an option in which the op id policy is ignored if the tserver is running out of disk space. If the gflag `log_stop_retaining_min_disk_mb` is set, we will stop retaining records with unreplicated entries once disk space is less than the specified value. This gflag will not affect log segments whose entries haven’t been flushed to rocksDB.
```